pkg/cluster/px/px.go

```diff
@@ -243,3 +256,217 @@ func (ops *pxClusterOps) runDockerPuller(imageToPull string) error {
+		deploymentName := d.Name
+		deploymentNamespace := d.Namespace
+		logrus.Infof("scaling down deployment: [%s] %s", deploymentNamespace, deploymentName)
+
+		t := func() (interface{}, bool, error) {
+			dCopy, err := ops.kubeClient.Apps().Deployments(deploymentNamespace).Get(deploymentName, metav1.GetOptions{})
+			if err != nil {
+				if matched, _ := regexp.MatchString(".+ not found", err.Error()); matched {
+					return nil, false, nil // done as deployment is deleted
+				}
+
+				return nil, true, err
+			}
+
+			dCopy = dCopy.DeepCopy()
+			dCopy.Annotations[replicaMemoryKey] = fmt.Sprintf("%d", *dCopy.Spec.Replicas)
+			dCopy.Spec.Replicas = &valZero
+
+			_, err = ops.kubeClient.Apps().Deployments(dCopy.Namespace).Update(dCopy)
+			if err != nil {
+				return nil, true, err
+			}
+
+			return nil, false, nil
+		}
+
+		if _, err := task.DoRetryWithTimeout(t, deploymentUpdateTimeout, defaultRetryInterval); err != nil {
+			return err
+		}
+	}
+
+	for _, s := range sharedSS {
+		logrus.Infof("scaling down statefulset: [%s] %s", s.Namespace, s.Name)
+
+		t := func() (interface{}, bool, error) {
+			sCopy, err := ops.kubeClient.Apps().StatefulSets(s.Namespace).Get(s.Name, metav1.GetOptions{})
+			if err != nil {
+				if matched, _ := regexp.MatchString(".+ not found", err.Error()); matched {
+					return nil, false, nil // done as statefulset is deleted
+				}
+
+				return nil, true, err
+			}
+
+			sCopy = sCopy.DeepCopy()
+			sCopy.Annotations[replicaMemoryKey] = fmt.Sprintf("%d", *sCopy.Spec.Replicas)
+			sCopy.Spec.Replicas = &valZero
+
+			_, err = ops.kubeClient.Apps().StatefulSets(sCopy.Namespace).Update(sCopy)
+			if err != nil {
+				return nil, true, err
+			}
+
+			return nil, false, nil
+		}
+
+		if _, err := task.DoRetryWithTimeout(t, statefulSetUpdateTimeout, defaultRetryInterval); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (ops *pxClusterOps) restoreScaledAppsReplicas() error {
+	sharedDeps, sharedSS, err := ops.getPXSharedApps()
+	if err != nil {
+		return err
+	}
+
+	logrus.Infof("found %d deployments and %d statefulsets to restore", len(sharedDeps), len(sharedSS))
+
+	for _, d := range sharedDeps {
+		deploymentName := d.Name
+		deploymentNamespace := d.Namespace
+		logrus.Infof("restoring app: [%s] %s", d.Namespace, d.Name)
+		t := func() (interface{}, bool, error) {
+			dCopy, err := ops.kubeClient.Apps().Deployments(deploymentNamespace).Get(deploymentName, metav1.GetOptions{})
+			if err != nil {
+				if matched, _ := regexp.MatchString(".+ not found", err.Error()); matched {
+					return nil, false, nil // done as deployment is deleted
+				}
+
+				return nil, true, err
+			}
+
+			dCopy = dCopy.DeepCopy()
+			if dCopy.Annotations == nil {
+				return nil, false, nil // done as this is not an app we touched
+			}
+
+			val, present := dCopy.Annotations[replicaMemoryKey]
+			if !present || len(val) == 0 {
+				return nil, false, nil // done as this is not an app we touched
```

Comment:
May want to log if it is not present, just in case someone deleted it.

Reply:
done